src/Microsoft.AspNetCore.Dispatcher.Abstractions/EndpointDelegateFactory.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    /// <summary>
+    /// A delegate which attempts to create a <see cref="RequestDelegate"/> for the selected <see cref="Endpoint"/>.
+    /// </summary>
+    /// <param name="httpContext">The <see cref="HttpContext"/> associated with the current request.</param>
+    /// <param name="requestDelegate">The next middleware in the pipeline. Usage of optional.</param>
+    /// <param name="endpoint">The <see cref="Endpoint"/> selected by the dispatcher.</param>
+    /// <returns>
+    /// A <see cref="RequestDelegate"/> that invokes the operation represented by the <see cref="Endpoint"/>, or <c>null</c>.
+    /// </returns>
+    public delegate RequestDelegate EndpointDelegateFactory(HttpContext httpContext, RequestDelegate requestDelegate, Endpoint endpoint);
```

Comment:
@davidfowl - I know we wanted to stick to delegate/simple signatures but this is horrible - need some suggestions 😆 

One idea is to not pass the endpoint. Then this ends up being the same as one of the `Use(...)` signatures. Buuuuuuuuuuuttt we expecte you to need the endpoint.... hmmm 🤔

Reply:
Make it return the middleware signature instead Func\<RequestDelegate, RequestDelegate\> does ignore need ge httpContext immediately? Can turn return a the middleware signature passing in the endpoint only?

Reply:
The reason I passed in the HttpContext originally is that I think you want to get access to features for things like route values, but maybe not. I'll go with your suggestion, and we can see if it works out or not.

Reply:
Also services.